### PR TITLE
Intermediate results loss reporting for Calibrate. Option for intermediate result reporting every integration step. 

### PR DESCRIPTION
**Pull request by jClugstor**
Adds callbacks for the solve of Calibrate so that optimization loss values and intermediate solutions can be reported. Added an option for intermediate results to be reported on every integration step instead of being time based. Added dispatches to the `get_callback` function to specialize on the type of the route inside of the `OperationRequest`.

### CODE REVIEW

File: src/operations.jl [hunk 1]
```diff
@@ -346,6 +354,28 @@ end
...
+    end
+end
+
+# callback for Simulate requests
+function get_callback(o::OperationRequest, ::Type{Simulate})
+    DiscreteCallback((args...) -> true, IntermediateResults(o.id,every_iteration = true),
+                                                    save_positions = (false,false))
+end
+
+
```
Comment (joshday) [inline src/operations.jl:378]: Move this next to the callback struct

File: src/operations.jl [hunk 2]
```diff
@@ -346,6 +354,28 @@ end
...
+    function (p,lossval,ode_sol)  
+        param_dict = Dict(parameters(ode_sol.prob.f.sys) .=> ode_sol.prob.p)
+        state_dict = Dict([state => ode_sol[state] for state in states(ode_sol.prob.f.sys)])
+        publish_to_rabbitmq(; loss = lossval, sol_data = state_dict, params = param_dict, id=o.id)
+    end
+end
+
+# callback for Simulate requests
+function get_callback(o::OperationRequest, ::Type{Simulate})
+    DiscreteCallback((args...) -> true, IntermediateResults(o.id,every_iteration = true),
```
Comment (joshday) [inline src/operations.jl:374]: Replace `every_iteration = true` with `every=Dates.Second(0)`

File: src/operations.jl [hunk 3]
```diff
@@ -346,6 +354,28 @@ end
...
 # :simulate => Simulate, etc.
+
+function get_callback(o::OperationRequest)
+    optype = route2operation_type[o.route]
+    get_callback(o,optype)
+end
+
+# callback for Calibrate requests
+function get_callback(o::OperationRequest, ::Type{Calibrate})
+    function (p,lossval,ode_sol)
```
Comment (joshday) [inline src/operations.jl:365]: Doesn't this need to be wrapped in a `DiscreteCallback` (or similar) object?

File: src/operations.jl [hunk 4]
```diff
@@ -199,10 +207,12 @@ function Calibrate(o::OperationRequest)
     Calibrate(sys, o.timespan, priors, data, num_chains, num_iterations, calibrate_method, ode_method)
 end
 
+# Jadon Clugston: maybe we want separate callbacks for the optimizer and the ode solve?
```
Comment (joshday) [inline src/operations.jl:210]: This may have been a key part I was missing in the original request for intermediate results.  I think we only care about the optimizer

File: src/operations.jl (hunk 3 above)
Comment (jClugstor) [reply src/operations.jl:365]: The problem is that this is an Optimization problem, so I don't think the DifferentialEquations callbacks apply to those. I followed what https://docs.sciml.ai/Optimization/stable/API/solve/ has to say about callbacks for optimization problems. There is an issue in easymodelanalysis talking about a time based callback https://github.com/SciML/EasyModelAnalysis.jl/issues/158 which we could switch to if it gets implemented, but for now all we need is a function that has that particular signature.
